### PR TITLE
[Feature, Hardware] Enable SGLang on AMD GPUs via PyTorch for ROCm

### PR DESCRIPTION
**Pull request by HaiShaw**
## Motivation
- Enable SGLang on AMD GPUs

## Modifications
- Bypass `FlashInfer` backend untill it is available on AMD/ROCm
- Add proper fix for AMD FP8 `e4m3fnuz` to support Fused_MoE
- Dependency over `vLLM>=0.5.5`, I modified `pyproject.toml` just to confirm that it works up to 0.6.0 as well.
- Misc.

- TODO: follow-up to address one error (below) when `cuda-graph` is enabled.
```
File "/sglang/python/sglang/srt/layers/sampler.py", line 164, in top_k_top_p_min_p_sampling_from_probs_torch
    min_p_thresholds = probs_sort[:, 0] * min_ps
    TypeError: unsupported operand type(s) for *: 'Tensor' and 'NoneType' (where min_ps is None)
```

## How to run
- An example on one MI3xx (not performance benchmark)
```
root@x:/sglang# VLLM_MOE_PADDING=0 python -m sglang.bench_latency --batch-size 32 --input 1024 --output 8 --model dummy_half_grok1/ --tokenizer-path Xenova/grok-1-tokenizer --load-format dummy --tp 8 --quant fp8  --attention-backend triton --sampling-backend  pytorch --disable-cuda-graph
Warmup ...
Prefill. latency: 25.79838 s, throughput:   1270.16 token/s
Decode.  latency: 0.53607 s, throughput:     59.69 token/s
Decode.  latency: 0.31325 s, throughput:    102.15 token/s
Decode.  latency: 0.04105 s, throughput:    779.47 token/s
Decode.  latency: 0.04075 s, throughput:    785.31 token/s
Decode.  median latency: 0.17715 s, median throughput:    180.64 token/s
Total. latency: 26.730 s, throughput:   1230.70 token/s
Benchmark ...
Prefill. latency: 1.11868 s, throughput:  29291.72 token/s
Decode.  latency: 0.02593 s, throughput:   1234.10 token/s
Decode.  latency: 0.02575 s, throughput:   1242.87 token/s
Decode.  latency: 0.02559 s, throughput:   1250.46 token/s
Decode.  latency: 0.02646 s, throughput:   1209.23 token/s
Decode.  latency: 0.02574 s, throughput:   1243.42 token/s
Decode.  median latency: 0.02574 s, median throughput:   1243.15 token/s
Total. latency:  1.325 s, throughput:  24925.60 token/s
root@x:/sglang#
```

## Checklist

- [+] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [+] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [+] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@ dependencies = [
 srt = ["aiohttp", "decord", "fastapi", "hf_transfer", "huggingface_hub", "interegular",
        "packaging", "pillow", "psutil", "pydantic", "python-multipart",
        "torch", "torchao", "uvicorn", "uvloop", "zmq",
-       "vllm==0.5.5", "outlines>=0.0.44"]
+       "vllm==0.6.0", "outlines>=0.0.44"]
```
Comment (merrymercy) [inline python/pyproject.toml]: Since your test has passed, revert this for this PR.

File: python/sglang/srt/layers/activation.py [hunk 2]
```diff
@@ -27,41 +26,50 @@
...
-        output_shape = x.shape[:-1] + (d,)
-        out = torch.empty(output_shape, dtype=x.dtype, device=x.device)
-        if self.approximate == "tanh":
-            gelu_tanh_and_mul(x, out)
-        elif self.approximate == "none":
-            gelu_and_mul(x, out)
-        else:
-            raise RuntimeError("GeluAndMul only support tanh or none")
-        return out
+from vllm.utils import is_hip
```
Comment (merrymercy) [inline python/sglang/srt/layers/activation.py]: we are going to remove the hard dependency on vllm gradually, we can copy `is_hip` to `srt/utils.py`

File: python/sglang/srt/layers/activation.py [hunk 3]
```diff
@@ -27,41 +26,50 @@
...
+# ROCm: flashinfer available later 
+if is_hip():
+    from vllm.model_executor.layers.activation import (
+        SiluAndMul,
+        GeluAndMul,
+    )
+else:
+    from flashinfer.activation import gelu_and_mul, gelu_tanh_and_mul, silu_and_mul
+
+    class SiluAndMul(CustomOp):
```
Comment (merrymercy) [inline python/sglang/srt/layers/activation.py]: I want to avoid this big if/else and avoid putting `class ` definition under an indentation. How about this for conditional import?

```python
logger = logging.getLogger(__name__)

try:
    if is_hip():
        raise ImportError("FlashInfer is not supported on AMD GPUs.")
    from flashinfer.activation import gelu_and_mul, gelu_tanh_and_mul, silu_and_mul
except ImportError as e:
    logger.warning("FlashInfer is not available. Fallback to other kernel libraries. Message: {e}")
    from vllm._custom_ops import gelu_and_mul, gelu_tanh_and_mul, silu_and_mul
```

File: python/sglang/srt/layers/layernorm.py [hunk 4]
```diff
@@ -19,93 +19,101 @@
...
-from flashinfer.norm import (
-    fused_add_rmsnorm,
-    gemma_fused_add_rmsnorm,
-    gemma_rmsnorm,
-    rmsnorm,
-)
 from vllm.model_executor.custom_op import CustomOp
-
-
-class RMSNorm(CustomOp):
```
Comment (merrymercy) [inline python/sglang/srt/layers/layernorm.py:31]: do the same things as `activation.py`

File: scripts/deprecated/test_flashinfer.py [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
-)
-from flashinfer.decode import _grouped_size_compiled_for_decode_kernels
 
 from sglang.srt.layers.token_attention import token_attention_fwd
 from sglang.srt.layers.triton_attention.extend_attention import (
     extend_attention_fwd,
     redundant_attention,
 )
 
+from vllm.utils import is_hip
```
Comment (merrymercy) [inline scripts/deprecated/test_flashinfer.py]: revert the change to this file.

File: python/sglang/srt/layers/attention_backend.py [hunk 6]
```diff
@@ -28,6 +21,16 @@
 if TYPE_CHECKING:
     from sglang.srt.model_executor.model_runner import ModelRunner
 
+from vllm.utils import is_hip
+# ROCm: flashinfer available later
+if not is_hip():
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention_backend.py]:
```suggestion
from sglang.srt.utils import is_hip
if not is_hip():
    # ROCm: flashinfer available later
```

Comment (merrymercy) [reply python/sglang/srt/layers/attention_backend.py]: propagate the change to other files

**Review by merrymercy — CHANGES_REQUESTED** [review]
1. We deprecated `--disable-flashinfer --disable-flashinfer-sampling`. Please use `--attention-backend triton --sampling-backend pytorch` instead.
2. Alternatively, you can set these backends automatically here https://github.com/HaiShaw/sglang/blob/8715deff22727382ce6f74768213e3e72f413f71/python/sglang/srt/server_args.py#L155
```python
if is_hip():
    self.attention_backend = "triton"
    self.sampling_backend = "pytorch"
```
3. The sampler + cuda graph issue has been fixed by https://github.com/sgl-project/sglang/pull/1392. You can rebase and try it again.
```python -m sglang.bench_latency --model-path TinyLlama/TinyLlama-1.1B-Chat-v0.4 --attention-backend triton --sampling-backend  pytorch``` This runs correctly.

File: python/sglang/srt/layers/layernorm.py (hunk 4 above)
Comment (HaiShaw) [reply python/sglang/srt/layers/layernorm.py:31]: `layernorm.py`: things are a bit different -
`vllm._custom_ops` has not defined `gemma_fused_add_rmsnorm`, `gemma_rmsnorm` equivalent, and `rms_norm` API is different.
So can we keep its current change for now?

File: python/sglang/srt/server.py [hunk 7]
```diff
@@ -434,6 +434,8 @@ def _set_envs_and_config(server_args: ServerArgs):
             "at https://docs.flashinfer.ai/installation.html.",
         )
 
+    mp.set_start_method("spawn", force=True)
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: only do this for hip for now. We will figure out a better method of not using fork later.

File: python/sglang/srt/layers/layernorm.py (hunk 4 above)
Comment (merrymercy) [reply python/sglang/srt/layers/layernorm.py:31]: I want to avoid a big indentation for class definition. Can you try this?
I found this is more general, please apply the same to `activation.py` as well.

```python
class RMSNorm(CustomOp):
    def __init__(
        self,
        hidden_size: int,
        eps: float = 1e-6,
    ) -> None:
        super().__init__()
        self.weight = nn.Parameter(torch.ones(hidden_size))
        self.variance_epsilon = eps

    def forward_cuda(
        self,
        x: torch.Tensor,
        residual: Optional[torch.Tensor] = None,
    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        ....

    def forward_native(
        self,
        x: torch.Tensor,
        residual: Optional[torch.Tensor] = None,
    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        ....

if is_hip():
    logger.info("FlashInfer is not available on AMD GPUs. Fallback to other kernel libraries.")
    from vllm.model_executor.layers.layernorm import GemmaRMSNorm, RMSNorm
    RMSNorm.forward_cuda = RMSNorm.forward_cuda
```

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/layernorm.py (hunk 4 above)
Comment (HaiShaw) [reply python/sglang/srt/layers/layernorm.py:31]: This reorder works! Let me make the change :)

File: python/sglang/srt/layers/activation.py [hunk 8]
```diff
@@ -64,6 +69,13 @@ def forward_cuda(self, x: torch.Tensor) -> torch.Tensor:
         return out
 
 
+if is_hip():
```
Comment (merrymercy) [inline python/sglang/srt/layers/activation.py]: 1. move this to the end of this file
2. do you need to skip the import of flashinfer at the beginning of this file?

Comment (HaiShaw) [reply python/sglang/srt/layers/activation.py]: @merrymercy just done 1. it is fine to keep as it is for 2.

**Review by merrymercy — APPROVED** [review]

**Comment by rj-linqingxu** [comment]
rocm 6.1.2
Memory access fault by GPU node-1 (Agent handle: 0x5629173ccc50) on address 0x7fcbccfeb000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
root@7acc50a881cf:/usr/local/lib/python3.10/dist-packages# /usr/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '

**Comment by merrymercy** [comment]
@linqingxu  try `--disable-cuda-graph` for now? We are working on more fixes.
